ocaml/xapi/repository.ml

```diff
@@ -695,26 +755,50 @@ let apply_updates' ~__context ~host ~updates_info ~livepatches ~acc_rpm_updates
   in
   (* Update states in cache *)
   update_cache ~host ~failed_livepatches ;
-  List.iter
-    (fun g -> debug "mandatory pending_guidance: %s" (to_string g))
-    mandatory ;
-  let ops = get_ops_of_pending ~__context ~host ~kind:Mandatory in
-  set_pending_guidances ~ops ~coming:mandatory ;
+  (* Set pending guidance lists *)
+  let set_guidance ~kind ~coming =
+    List.iter
+      (fun g ->
+        debug "setting %s pending guidance: %s" (kind_to_string kind)
+          (to_string g)
+      )
+      coming ;
+    let ops = get_ops_of_pending ~__context ~host ~kind in
+    set_pending_guidances ~ops ~coming
+  in
+  guidance |> List.iter (fun (kind, coming) -> set_guidance ~kind ~coming) ;
+  let get_livepatch_component (lp, _) = lp.LivePatch.component in
+  let applied = List.map get_livepatch_component applied_livepatches in
+  let failed = List.map get_livepatch_component failed_livepatches in
+  update_livepatch_failures ~__context ~host ~applied ~failed ;
   List.map
     (fun (lp, _) -> [Api_errors.apply_livepatch_failed; LivePatch.to_string lp])
     failed_livepatches
 
 let apply_updates ~__context ~host ~hash =
   (* This function runs on coordinator host *)
+  assert_no_host_pending_mandatory_guidance ~__context ~host ;
   try
     let repository = get_single_enabled_update_repository ~__context in
+    let repository_name = get_repository_name ~__context ~self:repository in
     if hash = "" || hash <> Db.Repository.get_hash ~__context ~self:repository
     then
       raise Api_errors.(Server_error (updateinfo_hash_mismatch, [])) ;
     with_pool_repositories (fun () ->
         let updates_info =
           parse_updateinfo ~__context ~self:repository ~check:true
         in
+        let updates_of_hosts =
+          if Helpers.is_pool_master ~__context ~host then (
+            (* save available updates before applying on coordinator *)
+            if Hashtbl.length updates_in_cache = 0 then
+              set_available_updates ~__context |> ignore ;
```

Comment:
Sorry, may I know why we need to `set_available_updates` here? Should it already be set inside the /updates API handler?

Reply:
The XAPI in coordinator may have restarted hence lost the cache.